File: packages/sdk-android/android/traceitx-core/src/main/kotlin/com/traceitx/envelope/InternalLogger.kt
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Process-wide ring buffer for txGuard{} failures. EnvelopeBuilder drains this
// into `captureControl.degradedReason` so the receiver knows the SDK degraded
// and can flag the report.
//
// Mirrors `packages/sdk-ios/Sources/TraceItX/Envelope/InternalLogger.swift`.
//
// Note: We use `android.util.Log` instead of Timber to avoid forcing Timber
// onto :traceitx-core's classpath (Timber is reflection-detected for log capture
// in Plan 05-04, NOT a hard dependency).
package com.traceitx.envelope

import java.time.Instant
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

@PublishedApi
internal object InternalLogger {
    private val lock = ReentrantLock()
    private val failures = ArrayDeque<Failure>()
    private const val MAX_BUFFERED = 50

    internal data class Failure(
        val label: String,
        val message: String,
        val at: Instant,
    )

    @PublishedApi
    internal fun recordSafeWrapFailure(label: String, error: Throwable) {
        lock.withLock {
            val entry = Failure(label, error.toString(), Instant.now())
            failures.addLast(entry)
            while (failures.size > MAX_BUFFERED) {
                failures.removeFirst()
            }
        }
        // Best-effort warn line; android.util.Log throws on JVM unit tests if not stubbed,
        // so route through System.err which Robolectric/JVM both tolerate.
        try {
            android.util.Log.w("TraceItX/safeWrap", "label=$label", error)
        } catch (_: Throwable) {
            System.err.println("TraceItX/safeWrap label=$label error=$error")
        }
    }

    /**
     * Drained by EnvelopeBuilder into `captureControl.degradedReason`. Buffer is
     * cleared after drain so each report only carries new failures.
     */
    fun drainFailures(): List<Failure> = lock.withLock {
        val out = failures.toList()
        failures.clear()
        out
    }
}

File: packages/sdk-ios/Sources/Everframe/Envelope/InternalLogger.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Process-wide ring buffer for safeWrap failures. EnvelopeBuilder drains this
// into `captureControl.degradedReason` so the receiver knows the Everframe SDK degraded
// and can flag the report. Public `os.Logger` line is also emitted at
// `.error` so customer-app crash console shows the failure path.
import Foundation
import os

enum InternalLogger {
    /// Logger + NSLock are Sendable in modern Foundation/os — let-bindings need
    /// no isolation modifier. `failures` is mutable global state, protected by
    /// `lock`; `nonisolated(unsafe)` stays only on the var.
    private static let log = Logger(subsystem: "dev.everframe", category: "internal")
    private static let lock = NSLock()
    nonisolated(unsafe) private static var failures: [Failure] = []
    private static let maxBuffered = 50

    struct Failure: Sendable {
        let label: String
        let message: String
        let at: Date
    }

    static func recordSafeWrapFailure(label: StaticString, error: Error) {
        lock.lock(); defer { lock.unlock() }
        let entry = Failure(label: "\(label)", message: "\(error)", at: Date())
        failures.append(entry)
        if failures.count > maxBuffered {
            failures.removeFirst(failures.count - maxBuffered)
        }
        log.error("safeWrap failed: \(entry.label, privacy: .public) — \(entry.message, privacy: .private)")
    }

    /// Drained by EnvelopeBuilder into `captureControl.degradedReason`. After
    /// drain the buffer is cleared so each report only carries new failures.
    @discardableResult
    static func drainFailures() -> [Failure] {
        lock.lock(); defer { lock.unlock() }
        let out = failures
        failures.removeAll()
        return out
    }
}
